lsp-imenu.el

```diff
@@ -43,32 +43,112 @@
+
+(defun lsp--symbol-get-start-point (sym)
+  "Get the start point of the name of SYM.
+
+SYM can be either DocumentSymbol or SymbolInformation."
+  (let* ((location (gethash "location" sym))
+         (name-range (or (and location (gethash "range" location))
+                         (gethash "selectionRange" sym)))
+         (start-point (lsp--position-to-point
+                       (gethash "start" name-range))))
+    (if imenu-use-markers (lsp--point-to-marker start-point) start-point)))
 
 (defun lsp--symbol-filter (sym)
-  (not
-    (lsp--equal-files
-      (lsp--uri-to-path (gethash "uri" (gethash "location" sym)))
-      (buffer-file-name))))
+  "Determine if SYM is for the current document."
+  (if-let ((location (gethash "location" sym)))
+      ;; It's a SymbolInformation
+      (not
+       (lsp--equal-files
+        (lsp--uri-to-path (gethash "uri" (gethash "location" sym)))
+        (buffer-file-name)))
+    ;; It's a DocumentSymbol, which is always in the current buffer file.
+    nil))
 
 (defun lsp--get-symbol-type (sym)
+  "The the string name of the kind of SYM."
   (or (cdr (assoc (gethash "kind" sym) lsp--symbol-kind)) "Other"))
 
 (defun lsp--imenu-create-index ()
-  (let ((symbols (seq-remove #'lsp--symbol-filter (lsp--get-document-symbols))))
-    (mapcar (lambda (nested-alist)
-              (cons (car nested-alist)
-                (mapcar #'lsp--symbol-to-imenu-elem (cdr nested-alist))))
-      (seq-group-by #'lsp--get-symbol-type symbols))))
+  "Create imenu index from document symbols."
+  (let ((symbols (lsp--get-document-symbols)))
+    (if (lsp--imenu-hierarchical-p symbols)
+        (lsp--imenu-create-hierarchical-index symbols)
+      (mapcar (lambda (nested-alist)
+                (cons (car nested-alist)
+                      (mapcar #'lsp--symbol-to-imenu-elem (cdr nested-alist))))
+              (seq-group-by #'lsp--get-symbol-type (lsp--imenu-filter-symbols symbols))))))
+
+(defun lsp--imenu-filter-symbols (symbols)
+  "Filter out unsupported symbols from SYMBOLS."
+  (seq-remove #'lsp--symbol-filter symbols))
+
+(defun lsp--imenu-hierarchical-p (symbols)
+  "Determine whether any element in SYMBOLS has children."
+  (seq-some (lambda (sym)
+              (gethash "children" sym))
+            symbols))
+
+(defun lsp--imenu-create-hierarchical-index (symbols)
+  "Create imenu index for hierarchical SYMBOLS.
+
+SYMBOLS are a list of DocumentSymbol messages.
+
+Return a nested alist keyed by symbol names. e.g.
+
+   ((\"SomeClass\" (\"(Class)\" . 10)
+                 (\"someField (Field)\" . 20)
+                 (\"someFunction (Function)\" . 25)
+                 (\"SomeSubClass\" (\"(Class)\" . 30)
+                                  (\"someSubField (Field)\" . 35))
+    (\"someFunction (Function)\" . 40))"
+  (let ((symbols (lsp--imenu-filter-symbols symbols)))
+    (mapcar (lambda (sym)
+              (lsp--symbol-to-hierarchical-imenu-elem sym))
+            (sort (lsp--imenu-filter-symbols symbols)
```

Comment:
I think in some cases sorting by position instead of name is better.

Reply:
Now it's customizable by `lsp-imenu-sort-methods`